crates/bevy_ui/src/widget/viewport.rs

```diff
@@ -0,0 +1,205 @@
+#[derive(Component, Debug, Clone, Copy, Reflect)]
+#[reflect(Component, Debug)]
+pub struct Viewport {
+    /// The entity representing the [`Camera`] associated with this viewport.
+    ///
+    /// Note that removing the [`Viewport`] component will not despawn this entity.
+    pub camera: Entity,
+}
+
+impl Viewport {
+    /// Creates a new [`Viewport`] with a given `camera`.
+    pub fn new(camera: Entity) -> Self {
+        Self { camera }
+    }
+}
+
+#[cfg(feature = "bevy_ui_picking_backend")]
+/// Handles viewport picking logic.
+///
+/// Viewport entities that are being hovered or dragged will have all pointer inputs sent to them.
+#[expect(
+    clippy::too_many_arguments,
+    reason = "System requires a lot of arguments"
+)]
+pub fn viewport_picking(
+    mut commands: Commands,
+    viewport_query: Query<(&Viewport, &PointerId, &Children)>,
+    node_query: Query<(&ComputedNode, &ImageNode, &GlobalTransform)>,
+    camera_query: Query<&Camera>,
+    hover_map: Res<HoverMap>,
+    pointer_state: Res<PointerState>,
+    mut pointer_inputs: EventReader<PointerInput>,
+    mut dragged_last_frame: Local<HashSet<(Entity, PointerId)>>,
+) {
+    let mut viewport_picks: HashSet<(Entity, PointerId)> = dragged_last_frame
+        .drain()
+        .chain(hover_map.iter().flat_map(|(hover_pointer_id, hits)| {
+            hits.iter()
+                .filter(|(entity, _)| viewport_query.contains(**entity))
+                .map(|(entity, _)| (*entity, *hover_pointer_id))
+        }))
+        .collect();
+
+    // Currently, we have only retrieved viewport entities if they are being hovered. However, this
+    // does not allow dragging in-and-out of viewports.
+    //
+    // We resolve this by considering viewports that are being dragged.
+    for ((pointer_id, _), pointer_state) in pointer_state.pointer_buttons.iter() {
+        for &target in pointer_state
+            .dragging
+            .keys()
+            .filter(|&entity| viewport_query.contains(*entity))
+        {
+            dragged_last_frame.insert((target, *pointer_id));
+            viewport_picks.insert((target, *pointer_id));
+        }
+    }
+
+    for (viewport_entity, pick_pointer_id) in viewport_picks {
+        let Ok((&viewport, &viewport_pointer_id, viewport_children)) =
+            viewport_query.get(viewport_entity)
+        else {
+            // This can only happen if entities in `dragged_last_frame` had one of these
+            // components removed since we last queried them
+            continue;
+        };
+
+        let Some((computed_node, image_node, global_transform)) = viewport_children
+            .iter()
+            .find_map(|child| node_query.get(*child).ok())
+        else {
+            continue;
+        };
+
+        let Some(cam_viewport_size) = camera_query
+            .get(viewport.camera)
+            .ok()
+            .and_then(Camera::logical_viewport_size)
+        else {
+            continue;
+        };
```

Comment:
```suggestion
        let Ok((&viewport, &viewport_pointer_id, computed_node, global_transform)) =
            viewport_query.get(viewport_entity)
        else {
            // This can only happen if entities in `dragged_last_frame` had one of these
            // components removed since we last queried them
            continue;
        };

        let Ok(camera) = camera_query.get(viewport.camera) else {
            continue;
        };

        let Some(cam_viewport_size) = camera.logical_viewport_size() else {
            continue;
        };
```